dataset/put-metadata.go

```diff
@@ -0,0 +1,95 @@
+		putMetadata(w, r, dc, zc, accessToken, collectionID, lang)
+	})
+}
+
+func putMetadata(w http.ResponseWriter, req *http.Request, dc DatasetClient, zc ZebedeeClient, userAccessToken, collectionID, lang string) {
+	ctx := req.Context()
+
+	if userAccessToken == "" {
+		log.Event(ctx, "no user access token header set", log.ERROR)
+		http.Error(w, "no user access token header set", http.StatusBadRequest)
+		return
+	}
+
+	if collectionID == "" {
+		log.Event(ctx, "no collection ID header set", log.ERROR)
+		http.Error(w, "no collection ID header set", http.StatusBadRequest)
+		return
+	}
+
+	vars := mux.Vars(req)
+	datasetID := vars["datasetID"]
+	edition := vars["editionID"]
+	version := vars["versionID"]
+
+	logInfo := map[string]interface{}{
+		"datasetID": datasetID,
+		"edition":   edition,
+		"version":   version,
+	}
+
+	b, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		log.Event(ctx, "putMetadata endpoint: error reading body", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error reading body", http.StatusBadRequest)
+		return
+	}
+
+	var body model.EditMetadata
+	if err = json.Unmarshal(b, &body); err != nil {
+		log.Event(ctx, "putMetadata endpoint: error unmarshalling body", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error unmarshalling body", http.StatusBadRequest)
+		return
+	}
+
+	err = dc.PutDataset(ctx, userAccessToken, "", collectionID, datasetID, body.Dataset)
+	if err != nil {
+		log.Event(ctx, "error updating dataset", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error updating dataset", http.StatusInternalServerError)
+		return
+	}
+
+	err = dc.PutVersion(ctx, userAccessToken, "", collectionID, datasetID, edition, version, body.Version)
+	if err != nil {
+		log.Event(ctx, "error updating version", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error updating version", http.StatusInternalServerError)
+		return
+	}
+
+	err = dc.PutInstance(ctx, userAccessToken, "", collectionID, datasetID, body.Instance)
+	if err != nil {
+		log.Event(ctx, "error updating dimensions", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error updating dimensions", http.StatusInternalServerError)
+		return
+	}
+
+	err = zc.PutDatasetInCollection(ctx, userAccessToken, "", collectionID, datasetID, body.CollectionState)
+	if err != nil {
+		log.Event(ctx, "error adding dataset to collection", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error adding dataset to collection", http.StatusInternalServerError)
+		return
+	}
+
+	err = zc.PutDatasetVersionInCollection(ctx, userAccessToken, "", collectionID, datasetID, edition, version, body.CollectionState)
+	if err != nil {
+		log.Event(ctx, "error adding version to collection", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error adding version to collection", http.StatusInternalServerError)
+	}
+
```

Comment:
Should we be setting a `200` or `204` status code in the responsewriter here?

Reply:
👍